### ISSUE TITLE
Can't find package denote

### ISSUE THREAD

**Issue by danielclucas** [radian-software/straight.el#997]
## What's wrong

I am trying to install the `denote` package through straight.el within Doom Emacs, but I keep getting "Could not find package denote." 

I have encountered the same error in a non-doom custom configuration where I use `use-package` statements to install packages, so I don't believe this is a doom problem. 

I have also run `straight-pull-recipe-repositories` so the mirrors should be up-to-date, and `denote` is in straight's elpa mirror. 

Any ideas on what might be causing this issue?

**Comment by progfolio** [comment]
Are you able to reproduce the issue via `straight-bug-report`?

**Comment by progfolio** [comment]
<details open><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :post-bootstrap 
  (straight-use-package 'denote))
```
</details>

- Test run at: `2022-09-06 00:40:45`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> develop, origin/develop) e20a44c 2022-09-01`
- emacs-version: `GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.34, cairo version 1.17.6)
 of 2022-09-01`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) e20a44c 2022-09-01
Cloning denote...
Cloning denote...done
Building denote...
Building denote...done

Packages:
"straight"                n/a                  develop e20a44c 2022-09-01
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master f0b8d824 2022-09-03
"gnu-elpa-mirror"         n/a                  master e195300 2022-09-03
"nongnu-elpa"             n/a                  main d52a19ccb0 2022-08-29
"el-get"                  melpa                master bf3dba44 2022-06-02
"emacsmirror-mirror"      n/a                  master d6b92f1 2022-08-29
"denote"                  gnu-elpa-mirror      master c69fb5c 2022-09-06


```
</details>

**Comment by danielclucas** [comment]
@progfolio I ran the test case 
```emacs-lisp
(straight-bug-report
  :post-bootstrap 
  (message "attempt to install denote")
  (straight-use-package
   '(denote))
  (message "bye"))
```
</details>

- Test run at: `2022-09-05 22:07:00`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> master) 4517e11 2022-04-13`
- emacs-version: `GNU Emacs 28.1 (build 2, x86_64-unknown-linux-gnu, GTK+ Version 3.24.31, cairo version 1.16.0)
 of 2022-04-16`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) e20a44c 2022-09-01
attempt to install denote
Cloning denote...
Cloning denote...done
Building denote...
Building denote...done

bye
Packages:
"straight"                n/a                  develop e20a44c 2022-09-01
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master f0b8d824 2022-09-03
"gnu-elpa-mirror"         n/a                  master e195300 2022-09-03
"nongnu-elpa"             n/a                  main d52a19ccb0 2022-08-29
"el-get"                  melpa                master bf3dba44 2022-06-02
"emacsmirror-mirror"      n/a                  master d6b92f1 2022-08-29
"denote"                  gnu-elpa-mirror      master c69fb5c 2022-09-06


```

As you can see, it ran successfully, but it took a very long time to run- at least 10 minutes. Seems like this might be a networking issue? What can I do at this point?

**Comment by progfolio** [comment]
Try the following test case.
It defaults to a shallow clone for the non-gnu-elpa repository.


<details open><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :pre-bootstrap 
  (setq straight-repository-user "progfolio" straight-repository-branch "fix/nongnu" start
        (current-time))
  :post-bootstrap 
  (straight-use-package 'denote)
  (message "elapsed time: %.2f seconds"
           (float-time
            (time-subtract
             (current-time)
             start))))
```
</details>

- Test run at: `2022-09-06 01:23:22`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> fix/nongnu, fork/fix/nongnu) 4e1e0aa 2022-09-06`
- emacs-version: `GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.34, cairo version 1.17.6)
 of 2022-09-01`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> fix/nongnu, origin/fix/nongnu) 4e1e0aa 2022-09-06
Cloning denote...
Cloning denote...done
Building denote...
Building denote...done

elapsed time: 16.27 seconds
Packages:
"straight"                n/a                  fix/nongnu 4e1e0aa 2022-09-06
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master f0b8d824 2022-09-03
"gnu-elpa-mirror"         n/a                  master e195300 2022-09-03
"nongnu-elpa"             n/a                  main d52a19c 2022-08-29
"el-get"                  melpa                master bf3dba44 2022-06-02
"emacsmirror-mirror"      n/a                  master d6b92f1 2022-08-29
"denote"                  gnu-elpa-mirror      master c69fb5c 2022-09-06


```
</details>

**Comment by danielclucas** [comment]
@progfolio Here's what I got. 
<details open><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :pre-bootstrap 
  (setq straight-repository-user "progfolio" straight-repository-branch "fix/nongnu" start
        (current-time))
  :post-bootstrap 
  (straight-use-package 'denote)
  (message "elapsed time: %.2f seconds"
           (float-time
            (time-subtract
             (current-time)
             start))))
```
</details>

- Test run at: `2022-09-05 22:46:49`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> master) 4517e11 2022-04-13`
- emacs-version: `GNU Emacs 28.1 (build 2, x86_64-unknown-linux-gnu, GTK+ Version 3.24.31, cairo version 1.16.0)
 of 2022-04-16`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...
Looking for nongnu-elpa recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...
Looking for emacsmirror-mirror recipe → Cloning nongnu-elpa...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> fix/nongnu, origin/fix/nongnu) 4e1e0aa 2022-09-06
Cloning denote...
Cloning denote...done
Building denote...
Building denote...done

elapsed time: 140.69 seconds
Packages:
"straight"                n/a                  fix/nongnu 4e1e0aa 2022-09-06
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master f0b8d824 2022-09-03
"gnu-elpa-mirror"         n/a                  master e195300 2022-09-03
"nongnu-elpa"             n/a                  main d52a19c 2022-08-29
"el-get"                  melpa                master bf3dba44 2022-06-02
"emacsmirror-mirror"      n/a                  master d6b92f1 2022-08-29
"denote"                  gnu-elpa-mirror      master c69fb5c 2022-09-06


```
</details>

This didn't take nearly as much time as the last test, though it still took some time. It wasn't the actual bootstrapping of straight.el thru to the building of denote that took the bulk of the time. It spent the most time, maybe a couple of minutes, hanging on the step `Looking for emacsmirror-mirror recipe`.

**Comment by progfolio** [comment]
I'm hesitant to add that as a workaround because it will cause problems with the lockfile system as described here:
 
https://github.com/radian-software/straight.el/issues/981

Implementing  https://github.com/radian-software/straight.el/issues/619 would make this an easy fix  by allowing users to remove the non-gnu-elpa recipe repository from straight-recipe-repositories.

Try timing a git clone of NonGNU ELPA's repo sans Emacs from the command line:

```bash
time git clone  "https://git.savannah.gnu.org/git/emacs/nongnu.git" /tmp/nongnu
```
For me that takes roughly 41 seconds. It would be even faster with a `--depth 1` option.

In either case, that will let you know if the networking issue is at all related to straight or if it's on NonGNU ELPA's end.
If it is on their end, the problem could be temporary.
If the issue persists, I will add a variable to exclude NonGNU ELPA from the recipe repository list as a workaround until #619 is implemented.

**Comment by danielclucas** [comment]
Running 
```shell
time git clone  "https://git.savannah.gnu.org/git/emacs/nongnu.git" /tmp/nongnu
```
Took me about 35 seconds. I then ran the original `straight-bug-report` code and it completed quickly. So it does seem that there was some sort of networking error that has now been resolved.

However, that still did not fix my ability to install denote. After removing my `.local` file and rebuilding everything from scratch it finally started working properly. I had done this before, but that hadn't fixed the issue. Perhaps it was the networking issue that prevented this from working originally. 

Anyway, thank you for your help.
